Fix persistent setup error notification content

frenck opened the pull request:
## Description:

Fixes the persistent setup error notifications contents:

- Does not replace underscores with hyphens in the name, since it makes it not true.
- Ensure the link points to the documentation location specified in the manifest.
- The error message still indicated `components`, changed to `integrations`

Result:

![image](https://user-images.githubusercontent.com/195327/70903849-17d2c800-2000-11ea-833b-040190167047.png)

~~Ideally, we should use the links from the manifest in the message, however, that is not available at that point. However, this takes care of 99% of the cases.~~

**Related issue (if applicable):** fixes #29884

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** n/a

## Example entry for `configuration.yaml` (if applicable):
```yaml
switch:
  - platform: wake_on_lan
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

frenck committed:
Fix persistent setup error notification content

MartinHjelmare commented:
Why isn't the link in the manifest available? The integration instance is passed to the function that calls this log function. The manifest is an attribute of the integration.

frenck commented:
In integration, the instance is not passed. Only the "component" name/string (which can be a platform as well).

I could use the loader to try to resolve it I guess? Is that what you are aiming for?

MartinHjelmare commented:
The integration instance is passes here:
https://github.com/home-assistant/home-assistant/blob/d851cb6f9e116dddb8ebeed4dd3910eb811993bd/homeassistant/config.py#L690-L692

It's this function that calls `async_log_exception`. So it would be easy to add access to the manifest. Just add a parameter.

frenck commented:
👍 Check, will adjust and update `_format_config_error` in that case as well, since that outputs now the link in a similar fashion as I had adjusted right now (but should be from the manifest as well in that case).

MartinHjelmare commented:
I did a search and we also use this log function from some other places, so it wasn't as clean as I first thought. We could default to `None` for `manfiest` parameter?

frenck commented:
We could simply pass the integration as suggested before. In case that is missing (defaulted to None), use the logic as suggested right now?

frenck commented:
Second thought, we could change the "link" parameter to be a string instead of a boolean. That way, we can decide on every location used, what to pass.

MartinHjelmare commented:
Changing the link parameter sound good!

frenck commented:
The problem becomes: The integration at hand, is the integration of the component entity at that point. So for the above example, it results in the manifest/link for the "switch" integration.

MartinHjelmare commented:
The platform integration is retrieved here:
https://github.com/home-assistant/home-assistant/blob/d851cb6f9e116dddb8ebeed4dd3910eb811993bd/homeassistant/config.py#L755

frenck committed:
Use documentation from manifest, enriched error messages

frenck commented:
Thanks for that hint @MartinHjelmare 👍 

I've adjusted that specific location. I now pushed the changes.
I've also added the documentation link into the error message at some additional points now. Since in some places, linking to the integration platform is actually really useful!

frenck committed:
Fix issue caught by mypy
